src/emailcontactstorage.cpp

```diff
@@ -0,0 +1,144 @@
+
+Q_GLOBAL_STATIC(EmailContactStorage, storage)
+EmailContactStorage* EmailContactStorage::instance()
+{
+    if (storage != nullptr && storage->m_isInit == false)
+        storage->initialize();
+    return storage;
+}
+
+
+void EmailContactStorage::initialize()
+{
+    QMailStore *mailStore = QMailStore::instance();
+    Q_ASSERT(mailStore);
+
+    if (QMailStore::initializationState() != QMailStore::Initialized)
+        return;
+
+    QMailMessageKey incomingKey(QMailMessageKey::status(QMailMessage::Incoming));
+    QMailMessageIdList incomingIds = mailStore->queryMessages(incomingKey);
+    for (const auto& id: incomingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QMailAddress from = metaData.from();
+        add(from, timeStamp);
+    }
+
+    QMailMessageKey outgoingKey(QMailMessageKey::status(QMailMessage::Outgoing));
+    QMailMessageIdList outgoingIds = mailStore->queryMessages(outgoingKey);
+    for (const auto& id: outgoingIds) {
+        const QMailMessageMetaData metaData = mailStore->messageMetaData(id);
+        if (metaData.messageType() != QMailMessageMetaDataFwd::MessageType::Email)
+            continue;
+
+        const QMailTimeStamp timeStamp = metaData.date();
+        const QList<QMailAddress> recipients = metaData.recipients();
+        for (const auto& address : recipients)
+            add(address, timeStamp);
+    }
+}
+
+
+int EmailContactStorage::count() const
+{
+    return m_storage.size();
+}
+
+
+QString EmailContactStorage::name(const int index) const
+{
+    return m_storage[index].name();
+}
+
+
+QString EmailContactStorage::email(const int index) const
+{
+    return m_storage[index].email();
+}
+
+
+EmailContact EmailContactStorage::at(const int index) const
+{
+    return m_storage[index];
+}
+
+// name can't have duplicate
+void EmailContactStorage::removeByName(const QString &name)
+{
+    auto iter = findByName(name);
+    if (iter == std::end(m_storage))
+        return;
+    m_storage.erase(iter);
+}
+
+// email can have duplicate. We must erase all email duplicates
+void EmailContactStorage::removeByEmail(const QString &email)
+{
+    for (auto iter = findByEmail(email);
+         iter != std::end(m_storage);
+         iter = findByEmail(email, iter, std::end(m_storage))) {
+        if (iter != std::end(m_storage))
+            iter = m_storage.erase(iter);
+    }
+}
+
+
+void EmailContactStorage::add(const QMailAddress &address, const QMailTimeStamp &stamp)
+{
+    if (!address.isEmailAddress())
+        return;
+
+    add(EmailContact(address.name(), address.address(), stamp));
+    add(EmailContact(address.address(), address.address(), stamp));
+}
+
+
+void EmailContactStorage::add(const EmailContact &contact)
+{
+    auto iter = findByName(contact.name());
```

Comment:
Now every insertion loops everything already inserted? Uh.

And why is it not allowing mutiple contacts by the same name? They could be even different persons.

Reply:
we often correspond with the same people. We will find matches very often. We need to replace the stamp with the latest one.

Reply:
the name here is the name of the EmailContact (it is not name of Contact).
This is what the user is finding (maybe a name or an address, what the user is finding we don't know). But the EmaiContact name is unique.